File: srpc/client-rpc.go
package srpc

import (
	"context"

	"github.com/pkg/errors"
)

// ClientRPC represents the client side of an on-going RPC call message stream.
type ClientRPC struct {
	commonRPC
}

// NewClientRPC constructs a call with the given service and method.
// Start attaches its transport; Close releases it.
func NewClientRPC(ctx context.Context, service, method string) *ClientRPC {
	rpc := &ClientRPC{}
	initCommonRPC(ctx, &rpc.commonRPC)
	rpc.service = service
	rpc.method = method
	return rpc
}

// Start attaches one writer and sends CallStart without holding the state lock.
func (r *ClientRPC) Start(writer PacketWriter, writeFirstMsg bool, firstMsg []byte) error {
	// Validate the call before transferring the transport to it.
	if writer == nil {
		return ErrNilWriter
	}
	if !writeFirstMsg {
		firstMsg = nil
	}
	pkt := NewCallStartPacket(r.service, r.method, firstMsg, writeFirstMsg && len(firstMsg) == 0)
	if err := pkt.Validate(); err != nil {
		return err
	}

	// Publish the writer before transport callbacks can deliver a response.
	locked := r.bcast.Lock()
	if r.writer != nil {
		locked.Unlock()
		return ErrCompleted
	}
	r.writer = writer
	locked.Broadcast()
	locked.Unlock()

	// Cancellation closes the transport even when its first write is blocked.
	stop := context.AfterFunc(r.ctx, r.Close)
	defer stop()
	if err := r.ctx.Err(); err != nil {
		r.Close()
		return err
	}
	if err := writer.WritePacket(pkt); err != nil {
		r.Close()
		return err
	}
	return nil
}

// HandlePacketData handles an incoming unparsed message packet.
func (r *ClientRPC) HandlePacketData(data []byte) error {
	pkt := &Packet{}
	if err := pkt.UnmarshalVT(data); err != nil {
		return err
	}
	return r.HandlePacket(pkt)
}

// HandleStreamClose handles the stream closing optionally w/ an error.
func (r *ClientRPC) HandleStreamClose(closeErr error) {
	r.commonRPC.HandleStreamClose(closeErr)
}

// HandlePacket handles an incoming parsed message packet.
func (r *ClientRPC) HandlePacket(msg *Packet) error {
	if err := msg.Validate(); err != nil {
		return err
	}

	switch b := msg.GetBody().(type) {
	case *Packet_CallStart:
		return r.HandleCallStart(b.CallStart)
	case *Packet_CallData:
		return r.HandleCallData(b.CallData)
	case *Packet_CallCancel:
		if b.CallCancel {
			return r.HandleCallCancel()
		}
		return nil
	default:
		return nil
	}
}

// HandleCallStart handles the call start packet.
func (r *ClientRPC) HandleCallStart(pkt *CallStart) error {
	return errors.Wrap(ErrUnrecognizedPacket, "call start packet unexpected")
}

// Close releases any resources held by the ClientRPC.
func (r *ClientRPC) Close() {
	// Settle the call before releasing a transport that can call back into it.
	locked := r.bcast.Lock()
	writer := r.closeLocked(&locked)
	locked.Unlock()

	// Closing the transport interrupts pending writes; sending cancellation first
	// could itself block forever behind a peer that stopped reading.
	if writer != nil {
		_ = writer.Close()
	}
}

File: srpc/client-rpc.cpp
//go:build deps_only

#include "client-rpc.hpp"

#include "rpcproto.pb.h"

namespace starpc {

ClientRPC::ClientRPC(const std::string &service, const std::string &method) {
  service_ = service;
  method_ = method;
}

ClientRPC::~ClientRPC() = default;

Error ClientRPC::Start(PacketWriter *writer, bool write_first_msg, const std::string &first_msg) {
  if (writer == nullptr)
    return Error::NilWriter;
  if (service_.empty())
    return Error::EmptyServiceID;
  if (method_.empty())
    return Error::EmptyMethodID;

  std::unique_lock writing(write_mutex_);
  {
    std::lock_guard lock(state_mutex_);
    if (writer_ != nullptr)
      return Error::Completed;
    writer_ = writer;
    if (canceled_ || data_closed_) {
      writing.unlock();
      // The caller closes a writer rejected before startup.
      return Error::Canceled;
    }
  }
  const auto packet = NewCallStartPacket(service_, method_, write_first_msg ? first_msg : "",
                                         write_first_msg && first_msg.empty());
  const auto written = writer->WritePacket(*packet);
  writing.unlock();
  if (written != Error::OK)
    HandleStreamClose(written);
  return written;
}

Error ClientRPC::HandlePacketData(const std::string &data) {
  srpc::Packet packet;
  if (!packet.ParseFromString(data))
    return Error::InvalidMessage;
  return HandlePacket(packet);
}

void ClientRPC::HandleStreamClose(Error close_err) { CommonRPC::HandleStreamClose(close_err); }

Error ClientRPC::HandlePacket(const srpc::Packet &packet) {
  const auto err = ValidatePacket(packet);
  if (err != Error::OK)
    return err;
  switch (packet.body_case()) {
  case srpc::Packet::kCallStart:
    return Error::UnrecognizedPacket;
  case srpc::Packet::kCallData:
    return HandleCallData(packet.call_data());
  case srpc::Packet::kCallCancel:
    return packet.call_cancel() ? HandleCallCancel() : Error::OK;
  default:
    return Error::UnrecognizedPacket;
  }
}

Error ClientRPC::HandleCallStart(const srpc::CallStart &) { return Error::UnrecognizedPacket; }

void ClientRPC::Close() { Cancel(); }

} // namespace starpc
